doc/regexp.rdoc

```diff
@@ -1,753 +1,1176 @@
+
+By default, a regexp is case-sensitive:
+
+  /foo/.match('FOO')  # => nil
+
+Modifier +i+ enables case-insensitive mode:
+
+  /foo/i.match('FOO')
+  # => #<MatchData "FOO">
+
+\Method Regexp#casefold? returns whether the mode is case-insensitive.
+
+=== Multiline Mode
+
+By default, a Ruby regexp is in multiline mode,
+which means that anchors <tt>^</tt> and <tt>$</tt> match, respectively,
+the beginning or end of a line in the target string.
+
+Modifier +m+ therefore does not enable multiline mode -- it's already enabled.
+The modifier does, however, make one small difference:
+
+- Without the +m+ modifier, a newline does not match subexpression <tt>.</tt>:
+
+    /a.c/.match("a\nc")  # => nil
+
+- With the modifier, it does match:
+
+    /a.c/m.match("a\nc") # => #<MatchData "a\nc">
+
+=== Extended Mode
+
+Modifier +x+ enables extended mode, which means that:
+
+- Literal white space in the pattern is to be ignored.
+- Character <tt>#</tt> marks the remainder of its containing line as a comment,
+  which is also to be ignored.
+
+In extended mode, whitespace and comments may be used
+to form a self-documented regexp.
+
+Regexp not in extended mode (matches some Roman numerals):
+
+  pattern = '^M{0,3}(CM|CD|D?C{0,3})(XC|XL|L?X{0,3})(IX|IV|V?I{0,3})$'
+  re = /#{pattern}/
+  re.match('MCMXLIII') # => #<MatchData "MCMXLIII" 1:"CM" 2:"XL" 3:"III">
+
+Regexp in extended mode:
+
+  pattern = <<-EOT
+    ^                   # beginning of string
+    M{0,3}              # thousands - 0 to 3 Ms
+    (CM|CD|D?C{0,3})    # hundreds - 900 (CM), 400 (CD), 0-300 (0 to 3 Cs),
+                        #            or 500-800 (D, followed by 0 to 3 Cs)
+    (XC|XL|L?X{0,3})    # tens - 90 (XC), 40 (XL), 0-30 (0 to 3 Xs),
+                        #        or 50-80 (L, followed by 0 to 3 Xs)
+    (IX|IV|V?I{0,3})    # ones - 9 (IX), 4 (IV), 0-3 (0 to 3 Is),
+                        #        or 5-8 (V, followed by 0 to 3 Is)
+    $                   # end of string
+  EOT
+  re = /#{pattern}/x
+  re.match('MCMXLIII') # => #<MatchData "MCMXLIII" 1:"CM" 2:"XL" 3:"III">
+
+=== Interpolation Mode
+
+By default, a regexp may be interplated repeatedly;
+in the +each+ loop below, the interpolated method call to +letters+
+is called three times, once for each loop iteration,
+and so takes more than 15 seconds to execute:
+
+  def letters; sleep 5; /[A-Z][a-z]/; end
+  words = %w[abc def xyz]
+  start = Time.now
+  words.each {|word| word.match(/\A[#{letters}]+\z/) }
+  Time.now - start # => 15.0174892
+
+With modifier +o+, the interpolated method is called only once,
+and so takes about 5 seconds to execute:
+
+  start = Time.now
+  words.each {|word| word.match(/\A[#{letters}]+\z/o) }
+  Time.now - start # => 5.0010866
+
+== Encodings
+
+By default, a regexp has US-ASCII encoding:
+
+  re = /foo/
+  re.source.encoding # => #<Encoding:US-ASCII>
+  re.encoding        # => #<Encoding:US-ASCII>
 
 Regular expressions are assumed to use the source encoding. This can be
```

Comment:
`Regular expressions containing non-US-ASCII characters are assumed to use the source encoding`

Reply:
Fixed.